examples_tests/23.Autoexposure/main.cpp

```diff
@@ -4,20 +4,771 @@
+
+				write.info = &info;
+
+				logicalDevice->updateDescriptorSets(1u, &write, 0u, nullptr);
+
+			}
+		}
+
+		// create FST graphics pipline
+		{
+			nbl::video::IGPUGraphicsPipeline::SCreationParams creationParams = {};
+			creationParams.renderpassIndependent = core::smart_refctd_ptr<nbl::video::IGPURenderpassIndependentPipeline>(const_cast<video::IGPURenderpassIndependentPipeline*>(fstRenderpassIndep.get()));
+			creationParams.renderpass = core::smart_refctd_ptr(renderpass);
+
+			fstGraphicsPipeline = logicalDevice->createGPUGraphicsPipeline(nullptr, std::move(creationParams));
+		}
+
+		for (uint32_t i = 0u; i < FRAMES_IN_FLIGHT; i++)
+		{
+			imageAcquire[i] = logicalDevice->createSemaphore();
+			renderFinished[i] = logicalDevice->createSemaphore();
+		}
+
+		lastPresentStamp = std::chrono::high_resolution_clock::now();
+	}
+
+	void onAppTerminated_impl() override
+	{
+		logicalDevice->waitIdle();
+	}
+
+	void workLoopBody() override
+	{
+		++resourceIx;
+		if (resourceIx >= FRAMES_IN_FLIGHT)
+			resourceIx = 0;
+
+		auto& cb = cmdbufs[resourceIx];
+		auto& fence = frameComplete[resourceIx];
+
+		if (fence)
+		{
+			logicalDevice->blockForFences(1u, &fence.get(), true);
+			logicalDevice->resetFences(1u, &fence.get());
+		}
+		else
+			fence = logicalDevice->createFence(static_cast<video::IGPUFence::E_CREATE_FLAGS>(0));
+
+		cb->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+		asset::SViewport viewport;
+		viewport.minDepth = 1.f;
+		viewport.maxDepth = 0.f;
+		viewport.x = 0u;
+		viewport.y = 0u;
+		viewport.width = WIN_W;
+		viewport.height = WIN_H;
+		cb->setViewport(0u, 1u, &viewport);
+
+		VkRect2D scissor;
+		scissor.offset = { 0, 0 };
+		scissor.extent = { WIN_W, WIN_H };
+		cb->setScissor(0u, 1u, &scissor);
+
+		swapchain->acquireNextImage(MAX_TIMEOUT, imageAcquire[resourceIx].get(), nullptr, &acquiredNextFBO);
+		cb->pushConstants(lumaMeteringPipeline->getLayout(), asset::IShader::ESS_COMPUTE, 0u, sizeof(outBufferIx[acquiredNextFBO]), &outBufferIx[acquiredNextFBO]);
+		outBufferIx[acquiredNextFBO] ^= 0x1u;
+		cb->bindDescriptorSets(asset::EPBP_COMPUTE, lumaMeteringPipeline->getLayout(), 0u, 1u, &lumaMeteringDS[acquiredNextFBO].get());
+		cb->bindComputePipeline(lumaMeteringPipeline.get());
+
+		video::IGPUCommandBuffer::SBufferMemoryBarrier lumaOutputBufferBarrier = {};
+		{
+			lumaOutputBufferBarrier.barrier.srcAccessMask = static_cast<asset::E_ACCESS_FLAGS>(0u);
+			lumaOutputBufferBarrier.barrier.dstAccessMask = asset::EAF_SHADER_READ_BIT;
+			lumaOutputBufferBarrier.srcQueueFamilyIndex = ~0u;
+			lumaOutputBufferBarrier.dstQueueFamilyIndex = ~0u;
+			lumaOutputBufferBarrier.buffer = toneParamsLumaOutputBuffer[acquiredNextFBO];
+			lumaOutputBufferBarrier.offset = sizeof(ToneMapperClass::Params_t<TMO>);
+			lumaOutputBufferBarrier.size = toneParamsLumaOutputBuffer[acquiredNextFBO]->getCachedCreationParams().declaredSize - lumaOutputBufferBarrier.offset;
```

Comment:
@Erfan-Ahmadi does Vulkan place some constraints on alignments and stuff on the size and offset? (smth like nonCoherentSize)